src/ILLink.RoslynAnalyzer/XmlAnalyzer.cs

```diff
@@ -0,0 +1,113 @@
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (s_moreThanOneValueForParameterOfMethod, s_errorProcessingXmlLocation);
+
+		private static readonly Regex _linkAttributesRegex = new (@"ILLink\.LinkAttributes.*\.xml");
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.EnableConcurrentExecution ();
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+			context.RegisterCompilationStartAction (context => {
+				// Report Diagnostics on malformed XML with additionalFileContext and register actions to resolve names with (CompilationStartAnalysisContext) context.
+				context.RegisterAdditionalFileAction (additionalFileContext => {
+					// Check if it's a LinkAttributes xml
+					if (_linkAttributesRegex.IsMatch (Path.GetFileName (additionalFileContext.AdditionalFile.Path))) {
+						if (additionalFileContext.AdditionalFile.GetText () is not SourceText text)
+							return;
+
+						if (!ValidateLinkAttributesXml (additionalFileContext, text))
+							return;
+
+						if (!context.TryGetValue (text, ProcessXmlProvider, out var xmlData) || xmlData is null) {
+							additionalFileContext.ReportDiagnostic (Diagnostic.Create (s_errorProcessingXmlLocation, null, additionalFileContext.AdditionalFile.Path));
+							return;
+						}
+						foreach (var root in xmlData) {
+							if (root is LinkAttributes.TypeNode typeNode) {
+								foreach (var duplicatedMethods in typeNode.Methods.GroupBy (m => m.Name).Where (m => m.Count () > 0)) {
+									additionalFileContext.ReportDiagnostic (Diagnostic.Create (s_moreThanOneValueForParameterOfMethod, null, duplicatedMethods.FirstOrDefault ().Name, typeNode.FullName));
+								}
+							}
+						}
+					}
+				});
+			});
+		}
+
+		private static XmlSchema GenerateLinkAttributesSchema ()
+		{
+			var assembly = Assembly.GetExecutingAssembly ();
+			using var schemaStream = assembly.GetManifestResourceStream ("ILLink.RoslynAnalyzer.ILLink.LinkAttributes.xsd");
+			using var reader = XmlReader.Create (schemaStream);
+			var schema = XmlSchema.Read (
+				reader,
+				null);
+			return schema;
+		}
+		private static readonly XmlSchema LinkAttributesSchema = GenerateLinkAttributesSchema ();
+
+		private static bool ValidateLinkAttributesXml (AdditionalFileAnalysisContext context, SourceText text)
+		{
+			var xmlStream = GenerateStream (text);
+			XDocument? document;
+			try {
+				document = XDocument.Load (xmlStream, LoadOptions.SetLineInfo | LoadOptions.PreserveWhitespace);
+			} catch (XmlException ex) {
+				context.ReportDiagnostic (Diagnostic.Create (s_errorProcessingXmlLocation, null, ex.Message));
+				return false;
+			}
+			XmlSchemaSet schemaSet = new XmlSchemaSet ();
+			schemaSet.Add (LinkAttributesSchema);
+			bool valid = true;
+			document.Validate (schemaSet, (sender, error) => { });
```

Comment:
If I read this correctly there will be no reported warning if the XSD validation fails - so what's the point of doing it?